src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/BookBinder.java

```diff
@@ -0,0 +1,138 @@
+                
+                if (enchantments.size() > 0) {
+                    ItemStack book = new ItemStack(Material.ENCHANTED_BOOK);
+                    book.setAmount(1);
+
+                    EnchantmentStorageMeta enchantMeta = (EnchantmentStorageMeta) book.getItemMeta();
+                    for (Map.Entry<Enchantment, Integer> entry : enchantments.entrySet()) {
+                        if (bypassVanillaMaxLevel.getValue()) {
+                            enchantMeta.addStoredEnchant(entry.getKey(), entry.getValue(), true);
+                        } else {
+                            enchantMeta.addStoredEnchant(entry.getKey(), entry.getValue(), false);
+                        }
+                    }
+
+                    book.setItemMeta(enchantMeta);
+                    
+                    MachineRecipe recipe = new MachineRecipe(25 * enchantments.size() / this.getSpeed(), new ItemStack[] {target, item}, new ItemStack[] {book});
+
+                    if (!InvUtils.fitAll(menu.toInventory(), recipe.getOutput(), getOutputSlots())) {
+                        return null;
+                    }
+
+                    for (int inputSlot : getInputSlots()) {
+                        menu.consumeItem(inputSlot);
+                    }
+
+                    return recipe;
+                }
+
+                return null;
+            }
+
+        }
+
+        return null;
+    }
+
+    private boolean isCompatible(@Nullable ItemStack item) {
+        if (item != null && item.getType() == Material.ENCHANTED_BOOK) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public ItemStack getProgressBar() {
+        return new ItemStack(Material.IRON_CHESTPLATE);
+    }
+
+    @Override
+    public String getMachineIdentifier() {
+        return "BOOK_BINDER";
+    }
+
+    private Map<Enchantment, Integer> combineEnchantments(Map<Enchantment, Integer> ech1, Map<Enchantment, Integer> ech2) {
+        Map<Enchantment, Integer> enchantments = new HashMap<>();
+        
+        enchantments.putAll(ech1);
+        for (Map.Entry<Enchantment, Integer> entry : ech2.entrySet()) {
+            enchantments.merge(entry.getKey(), entry.getValue(), (a, b) -> {
+                    if (a == b) {
+                        if (hasCustomMaxLevel.getValue()) {
+                            return a + 1 > customMaxLevel.getValue() ? customMaxLevel.getValue() : a + 1;
+                        } else {
+                            return a + 1;
+                        }
+                        
+                    } else {
+                        if (hasCustomMaxLevel.getValue()) {
+                            return Math.max(a, b) > customMaxLevel.getValue() ? customMaxLevel.getValue() : Math.max(a, b);
+                        } else {
+                            return Math.max(a, b);
+                        }
```

Comment:
You are doing Math.max(a,b) here regardless of the state, in one branch even twice.
So I think it would be wise to create a variable here.

```suggestion
                        int highestLevel = Math.max(a, b);
                        if (hasCustomMaxLevel.getValue()) {
                            return highestLevel > customMaxLevel.getValue() ? customMaxLevel.getValue() : highestLevel;
                        } else {
                            return highestLevel;
                        }
```